ext/opentelemetry-ext-asyncpg/src/opentelemetry/ext/asyncpg/__init__.py

```diff
@@ -0,0 +1,146 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This library allows tracing PostgreSQL queries made by the
+`asyncpg <https://magicstack.github.io/asyncpg/current/>`_ library.
+
+Usage
+-----
+
+.. code-block:: python
+
+    import asyncpg
+    import opentelemetry.ext.asyncpg
+
+    # You can optionally pass a custom TracerProvider to AsyncPGInstrumentor.instrument()
+    opentelemetry.ext.asyncpg.AsyncPGInstrumentor().instrument()
+    conn = await asyncpg.connect(user='user', password='password',
+                                 database='database', host='127.0.0.1')
+    values = await conn.fetch('''SELECT 42;''')
+
+API
+---
+"""
+
+import functools
+
+from asyncpg import Connection, exceptions
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.trace import SpanKind
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+_APPLIED = "_opentelemetry_ext_asyncpg_applied"
+
+
+def _exception_to_canonical_code(exc: Exception) -> StatusCanonicalCode:
+    if isinstance(exc, (exceptions.InterfaceError,),):
+        return StatusCanonicalCode.INVALID_ARGUMENT
+    if isinstance(exc, exceptions.IdleInTransactionSessionTimeoutError):
+        return StatusCanonicalCode.DEADLINE_EXCEEDED
+    return StatusCanonicalCode.UNKNOWN
+
+
+def _hydrate_span_from_args(span, *args, **__):
+    span.set_attribute("db.type", "sql")
+
+    if len(args) <= 0:
+        return span
+
+    connection = args[0]
+    if connection is not None:
+        params = getattr(connection, "_params", None)
+        if params is not None:
+            database_name = getattr(params, "database", None)
+            if database_name is not None:
+                span.set_attribute("db.instance", database_name)
+
+            database_user = getattr(params, "user", None)
+            if database_user is not None:
+                span.set_attribute("db.user", database_user)
```

Comment:
Is there a guarantee that if one of these parameters is on the connection then others will be? or do we actually need to check if each one is not none

Reply:
fixed for to much simpler params extraction